### DISCUSSION TITLE
Intra-document masking - cross contamination and distractions during pre-training

### DISCUSSION THREAD

**Discussion by biuq** [karpathy/llm.c#690] [Ideas]
Hello!

According to the paper [Analysing The Impact of Sequence Composition on Language Model Pre-Training](https://arxiv.org/abs/2402.13991), it seems that instead of letting the model learn document boundaries with EOS/EOT tokens, it's "better" to prevent the attention from crossing document/sequence boundaries. This was also the method [used in LLaMa 3 models](https://ai.meta.com/blog/meta-llama-3/):

> We trained the models on sequences of 8,192 tokens, using a mask to ensure self-attention does not cross document boundaries.

I wonder whether this idea has already been explored for this repo?
